src/main/java/org/tugraz/sysds/runtime/data/HeterogTensor.java

```diff
@@ -0,0 +1,383 @@
+			_schema = Arrays.copyOfRange(_schema, 0, getDim(1));
+		if (_colsdata == null) {
+			allocateBlock();
+		} else {
+			HomogTensor[] newCols = new HomogTensor[getDim(1)];
+			if (_colsdata.length > getDim(1))
+				_colsdata = Arrays.copyOfRange(_colsdata, 0, getDim(1));
+			int[] blockDims = toInternalDims(dims);
+			for (HomogTensor colsdata : _colsdata) {
+				colsdata.reset(blockDims);
+			}
+		}
+	}
+
+	public void reset(int[] dims, ValueType[] schema) {
+		if (dims.length < 2)
+			throw new DMLRuntimeException("HeterogTensor.reset(int[],ValueType[]) invalid number of tensor dimensions: " + dims.length);
+		if (dims[1] != schema.length)
+			throw new DMLRuntimeException("HeterogTensor.reset(int[],ValueType[]) column dimension and schema length does not match");
+		for (int i = 0; i < dims.length; i++)
+			if (dims[i] < 0)
+				throw new DMLRuntimeException("HeterogTensor.reset(int[],ValueType[]) invalid  dimension " + i + ": " + dims[i]);
+		_dims = dims;
+		_schema = schema;
+		if (_colsdata == null) {
+			allocateBlock();
+		} else {
+			HomogTensor[] newCols = new HomogTensor[getDim(1)];
+			int[] blockDims = toInternalDims(dims);
+			for (int c = 0; c < getDim(1); c++) {
+				if (_colsdata[c]._vt != schema[c]) {
+					newCols[c] = new HomogTensor(schema[c], blockDims, false);
+				}
+				else {
+					newCols[c] = _colsdata[c];
+				}
+				newCols[c].reset(blockDims);
+			}
+			_colsdata = newCols;
+		}
+	}
+
+	@Override
+	public HeterogTensor allocateBlock() {
+		if (_colsdata == null)
+			_colsdata = new HomogTensor[getDim(1)];
+		// All column tensors can share the same dimension array
+		// since their dimension should always be equal.
+		int[] blockDims = toInternalDims(_dims);
+		for (int i = 0; i < _schema.length; i++) {
+			// TODO sparse
+			_colsdata[i] = new HomogTensor(_schema[i], blockDims, false);
+			_colsdata[i].allocateBlock();
+		}
+		return this;
+	}
+
+	@Override
+	public boolean isAllocated() {
+		if (_colsdata == null)
+			return false;
+		for (HomogTensor block : _colsdata) {
+			if (block.isAllocated())
+				return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean isEmpty(boolean safe) {
+		if (!isAllocated()) {
+			return true;
+		}
+		for (HomogTensor tb : _colsdata) {
+			if (!tb.isEmpty(safe))
+				return false;
+		}
+		return true;
+	}
+
+	public ValueType[] getSchema() {
+		return _schema;
+	}
+
+	public ValueType getColValueType(int col) {
+		return _schema[col];
+	}
+
+	@Override
+	public double get(int[] ix) {
+		int[] internalIx = toInternalIx(ix);
+		return _colsdata[ix[1]].get(internalIx);
+	}
+
+	@Override
+	public double get(int r, int c) {
+		if (getNumDims() != 2)
+			throw new DMLRuntimeException("HeterogTensor.get(int,int) dimension mismatch: expected=2 actual=" + getNumDims());
+		return _colsdata[c].get(r, 0);
+	}
+
+	@Override
+	public long getLong(int[] ix) {
+		int[] internalIx = toInternalIx(ix);
+		return _colsdata[ix[1]].getLong(internalIx);
+	}
```

Comment:
At tensor block level, we could simply return objects and doubles - we only need the type specific methods on DenseBlock because all operations will be implemented against these blocks not against TensorBlock

Reply:
Could you explain what you mean by "objects and doubles"? Should we have two methods, one for Object (In which I would return either Long, Double, String etc. I guess) and one for returning doubles?